contrib/checkstyle/src/mill/contrib/checkstyle/CheckstyleModule.scala

```diff
@@ -0,0 +1,195 @@
+
+  import CheckstyleModule._
+
+  /**
+   * Generates a [[Checkstyle]].
+   */
+  def checkstyle: T[Checkstyle] = T {
+
+    val dest = T.dest
+
+    val format = checkstyleFormat()
+    val report = dest / s"checkstyle-report.$format"
+
+    val args = checkstyleOptions() ++
+      Seq(
+        "-c",
+        checkstyleConfig().path.toString(),
+        "-f",
+        format,
+        "-o",
+        report.toString()
+      ) ++
+      sources().map(_.path.toString())
+
+    T.log.info("generating checkstyle report ...")
+    T.log.debug(s"running checkstyle with $args")
+
+    val errors = Jvm.callSubprocess(
+      mainClass = "com.puppycrawl.tools.checkstyle.Main",
+      classPath = checkstyleClasspath().map(_.path),
+      mainArgs = args,
+      workingDir = dest,
+      check = false
+    ).exitCode
+
+    if (errors == 0) {
+      T.log.info(s"checkstyle found no errors, details in $report")
+    } else if (errors > 0 && os.exists(report)) {
+      T.log.error(s"checkstyle found $errors error(s), details in $report")
+    } else {
+      // errors is always 255, assuming byte value is the exit code
+      val exit = errors.toByte
+      T.log.error(
+        s"checkstyle exit($exit), please check plugin settings or try with another version"
+      )
+      throw new UnsupportedOperationException(s"checkstyle exit($exit)")
+    }
+    val transformations =
+      checkstyleTransformer(format).fold(Seq.empty[PathRef]) { transformer =>
+        val f = transformer(report)
+        checkstyleTransforms().map {
+          case (ref, rel) =>
+            val transform = ref.path
+            val out = dest / rel
+            T.log.info(s"transforming checkstyle report using $transform")
+            f(transform, out)
+            T.log.info(s"transformed checkstyle report to $out")
+            PathRef(out)
+        }.toSeq
+      }
+
+    Checkstyle(errors, PathRef(report), transformations)
+  }
+
+  /**
+   * Classpath for running Checkstyle.
+   */
+  def checkstyleClasspath: T[Loose.Agg[PathRef]] = T {
+    defaultResolver().resolveDeps(
+      Agg(ivy"com.puppycrawl.tools:checkstyle:${checkstyleVersion()}")
+    )
+  }
+
+  /**
+   * Checkstyle configuration file. Defaults to [[checkstyleDir]] `/` `config.xml`.
+   */
+  def checkstyleConfig: T[PathRef] = T.source {
+    checkstyleDir().path / "config.xml"
+  }
+
+  /**
+   * Directory containing Checkstyle configuration and transformations. Defaults to `checkstyle`.
+   */
+  def checkstyleDir: T[PathRef] = T {
+    PathRef(millSourcePath / "checkstyle")
+  }
+
+  /**
+   * Checkstyle report format. Defaults to `xml`.
+   */
+  def checkstyleFormat: T[String] = T {
+    "xml"
+  }
+
+  /**
+   * Additional arguments for Checkstyle.
+   *
+   * @see [[https://checkstyle.org/cmdline.html#Command_line_usage CLI options]]
+   */
+  def checkstyleOptions: T[Seq[String]] = T {
+    Seq.empty[String]
+  }
+
+  /**
+   * Returns a [[Transformer transformer]] for `format`, if supported.
```

Comment:
A bit more verbosity here would be useful. What is a `transformer`? What does it transform? What would I as a Java programmer who's not an expert in checkstyle care about this?

Reply:
Also, what's the difference between this and `def checkstyleTransform` below?

Reply:
Type aliases were extracted out to classes.